### PR TITLE
[SPARK-18123][SQL] Use db column names instead of RDD column ones during JDBC Writing

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Apache Spark supports the following cases **by quoting RDD column names** while saving through JDBC.
- Allow reserved keyword as a column name, e.g., 'order'.
- Allow mixed-case colume names like the following, e.g., `[a: int, A: int]`.
  
  ``` scala
  scala> val df = sql("select 1 a, 1 A")
  df: org.apache.spark.sql.DataFrame = [a: int, A: int]
  ...
  scala> df.write.mode("overwrite").format("jdbc").options(option).save()
  scala> df.write.mode("append").format("jdbc").options(option).save()
  ```

This PR aims to use **database column names** instead of RDD column ones in order to support the following additionally.
Note that this case succeeds with `MySQL`, but fails on `Postgres`/`Oracle` before.

``` scala
val df1 = sql("select 1 a")
val df2 = sql("select 1 A")
...
df1.write.mode("overwrite").format("jdbc").options(option).save()
df2.write.mode("append").format("jdbc").options(option).save()
```
## How was this patch tested?

Pass the Jenkins test with a new testcase.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -112,7 +112,25 @@ object JdbcUtils extends Logging {
    */
   def insertStatement(conn: Connection, table: String, rddSchema: StructType, dialect: JdbcDialect)
       : PreparedStatement = {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Actually, this is an approach similar to `normalizePartitionSpec` in `PartitioningUtils.scala`.

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L248-L251

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Two failures seems to be irrelevant. 
```
[info] - randomized aggregation test - [typed, with partial + safe] - with grouping keys - with non-empty input *** FAILED *** (1 second, 367 milliseconds)
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 2]
```diff
@@ -112,7 +112,25 @@ object JdbcUtils extends Logging {
...
+    var columns: String = ""
+    try {
+      val tableSchema = getSchema(tableSchemaQuery.executeQuery(), dialect)
+      val nameMap = tableSchema.fields.map(f => f.name -> f.name).toMap
+      val lowercaseNameMap = tableSchema.fields.map(f => f.name.toLowerCase -> f.name).toMap
+      columns = rddSchema.fields.map { x =>
+        if (nameMap.isDefinedAt(x.name)) {
+          dialect.quoteIdentifier(x.name)
+        } else if (lowercaseNameMap.isDefinedAt(x.name.toLowerCase)) {
+          dialect.quoteIdentifier(nameMap(x.name.toLowerCase))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: nameMap -> lowercaseNameMap.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @viirya .
I'll update like that.

**Comment by gatorsmile** [comment]
This is a bug fix, right? Will review this tomorrow.

**Comment by dongjoon-hyun** [comment]
Yes, right! Thank you, @gatorsmile !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 3]
```diff
@@ -112,7 +112,25 @@ object JdbcUtils extends Logging {
    */
   def insertStatement(conn: Connection, table: String, rddSchema: StructType, dialect: JdbcDialect)
       : PreparedStatement = {
-    val columns = rddSchema.fields.map(x => dialect.quoteIdentifier(x.name)).mkString(",")
+    // Use database column names instead of RDD schema column names
+    val tableSchemaQuery = conn.prepareStatement(dialect.getSchemaQuery(table))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I think the table schema won't change during inserting all data. You ask table schema for every insert statement now. Can we do this once in caller side (i.e., `savePartition`) and reuse the schema then?

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Thank you for review, @viirya . I'll try to update like that.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: We can get the table schema [when we checking whether the table exists](https://github.com/apache/spark/blob/fb07bbe575aabe68422fd3a31865101fb7fa1722/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala#L63).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 4]
```diff
@@ -112,7 +112,25 @@ object JdbcUtils extends Logging {
...
+      val nameMap = tableSchema.fields.map(f => f.name -> f.name).toMap
+      val lowercaseNameMap = tableSchema.fields.map(f => f.name.toLowerCase -> f.name).toMap
+      columns = rddSchema.fields.map { x =>
+        if (nameMap.isDefinedAt(x.name)) {
+          dialect.quoteIdentifier(x.name)
+        } else if (lowercaseNameMap.isDefinedAt(x.name.toLowerCase)) {
+          dialect.quoteIdentifier(lowercaseNameMap(x.name.toLowerCase))
+        } else {
+          throw new SQLException(s"""Column "${x.name}" not found""")
+        }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: The name resolution should be still controlled by `spark.sql.caseSensitive`, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 5]
```diff
@@ -112,7 +112,25 @@ object JdbcUtils extends Logging {
...
+      }.mkString(",")
+    } finally {
+      tableSchemaQuery.close()
+    }
     val placeholders = rddSchema.fields.map(_ => "?").mkString(",")
     val sql = s"INSERT INTO $table ($columns) VALUES ($placeholders)"
     conn.prepareStatement(sql)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Can we build the INSERT SQL statement in `saveTable` based on the schema? No need to prepare the generated statement in `saveTable`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 3 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Thank you for review, @gatorsmile .
Yes. That looks great! We can use `getSchemaQuery` instead of `tableExist`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 4 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yep. I'll fix that.

**Comment by dongjoon-hyun** [comment]
The PR is updated to 
- get table schema once in `createRelation`
- respect `spark.sql.caseSensitive`

For `insertStatement`, I thought it seems to be better to keep the current structure to isolate `INSERT` PreparedStatement generation code like `dialect.quoteIdentifier`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 6]
```diff
@@ -211,6 +211,52 @@ object JdbcUtils extends Logging {
...
+        getSchema(statement.executeQuery(), dialect)
+      } finally {
+        statement.close()
+      }
+    } match {
+      case Success(v) =>
+        Some(v)
+      case Failure(e) =>
+        None
+    }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Please do not use Try/Success/Failure. https://github.com/databricks/scala-style-guide#exception-handling-try-vs-try

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 7]
```diff
@@ -60,23 +60,27 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
+      val caseSensitive = sqlContext.conf.caseSensitiveAnalysis
       if (tableExists) {
         mode match {
           case SaveMode.Overwrite =>
-            if (isTruncate && isCascadingTruncateTable(url) == Some(false)) {
+            val savingSchema = if (isTruncate && isCascadingTruncateTable(url) == Some(false)) {
               // In this case, we should truncate table and then load.
               truncateTable(conn, table)
-              saveTable(df, url, table, jdbcOptions)
+              JdbcUtils.getSavingSchema(df.schema, tableSchema.get, caseSensitive)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Based on your current implementation, the `tableSchema` could be None, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 8]
```diff
@@ -211,6 +211,52 @@ object JdbcUtils extends Logging {
...
+  }
+
+  /**
+   * Returns the saving schema using rddSchema's sequence and tableSchema's name.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `saving schema` is not right. We need a better name here. 

`rddSchema's sequence and tableSchema's name` -> `rddSchema's column sequence and tableSchema's column names`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 9]
```diff
@@ -211,6 +211,52 @@ object JdbcUtils extends Logging {
...
+    val nameMap = tableSchema.fields.map(f => f.name -> f).toMap
+    val lowercaseNameMap = tableSchema.fields.map(f => f.name.toLowerCase -> f).toMap
+
+    var schema = new StructType()
+    rddSchema.fields.foreach { f =>
+      if (nameMap.isDefinedAt(f.name)) {
+        // identical names
+        schema = schema.add(nameMap(f.name))
+      } else if (!caseSensitive && lowercaseNameMap.isDefinedAt(f.name.toLowerCase)) {
+        // case-insensitive identical names
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Need to improve the comments. Actually, we return case sensitive column names.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 10]
```diff
@@ -211,6 +211,52 @@ object JdbcUtils extends Logging {
...
+        schema = schema.add(lowercaseNameMap(f.name.toLowerCase))
+      } else {
+        throw new org.apache.spark.SparkException(s"""Column "${f.name}" not found""")
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `org.apache.spark.SparkException ` -> `AnalysisException`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Here, we need to explain why we need to use the column sequences in `rddSchema` and why we need to use the column names in `tableSchema`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 11]
```diff
@@ -60,23 +60,27 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
             } else {
               // Otherwise, do not truncate the table, instead drop and recreate it
               dropTable(conn, table)
               createTable(df.schema, url, table, createTableOptions, conn)
-              saveTable(df, url, table, jdbcOptions)
+              df.schema
             }
+            saveTable(df, url, table, savingSchema, jdbcOptions)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: How about passing the table schema and resolve the schema inside `saveTable`? It might simplify the codes

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 6 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I see. Thank you for review! I wrote that by keeping the same logic in [tableExists](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L66-L80
) because I thought the guideline is about using return `Try` as a return value before this PR. 

Sure, I'll remove the usage of `Try/Success/Failure`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yep. I'll add more correct details here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 9 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: My bad. I meant `case-insensitively identical names`. I'll revise this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 10 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Sure!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 11 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: That could be. But, to do that, `JdbcUtil.saveTable` need to understand `SaveMode`, too. Is it okay?
Currently, `JdbcUtil` only provides somewhat primitive APIs.

**Comment by dongjoon-hyun** [comment]
The failure is not related to this PR.
```
[info] - fatal errors from a source should be sent to the user *** FAILED *** (84 milliseconds)
[info]   org.apache.spark.sql.streaming.StreamingQueryException: Query [id = 5ae8eeb2-81f1-44df-b8ad-55b780512ee5, runId = 3a5bfebc-e3a3-496e-987b-e0e789f1fdcb] terminated with exception: null
```

**Comment by gatorsmile** [comment]
You can move normalization of schemas into `saveTable` without passing save mode. See my private branch: https://github.com/apache/spark/compare/master...gatorsmile:pr-15664Changed?expand=1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 12]
```diff
@@ -211,6 +211,55 @@ object JdbcUtils extends Logging {
...
+   * Returns a schema using rddSchema's column sequence and tableSchema's column names.
+   *
+   * When appending data into some case-sensitive DBMSs like PostgreSQL/Oracle, we need to respect
+   * the existing case-sensitive column names instead of RDD column names for user convenience.
+   * See SPARK-18123 for more details.
+   */
+  def normalizeSchema(
+      rddSchema: StructType,
+      tableSchema: StructType,
+      caseSensitive: Boolean): StructType = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `isCaseSensitive`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 13]
```diff
@@ -211,6 +211,55 @@ object JdbcUtils extends Logging {
   }
 
   /**
+   * Returns the schema if the table already exists in the JDBC database.
+   */
+  def getSchema(conn: Connection, url: String, table: String): Option[StructType] = {
+    val dialect = JdbcDialects.get(url)
+
+    try {
+      val statement = conn.prepareStatement(dialect.getSchemaQuery(table))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: What is the reason why this should be included in the try block?

**Comment by dongjoon-hyun** [comment]
Wow. Thank you for the code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 13 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yes. It's due to `PreparedStatement prepareStatement(String sql) throws SQLException`. I want to make it sure to return `None` instead of raising SQLException at this line.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Is it better not to hide that here? If then, I'll remove that `try`.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: JDBC data sources might detect whether the table exists when preparing the SQL statement. Thus, please keep it. Thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 14]
```diff
@@ -568,10 +617,10 @@ object JdbcUtils extends Logging {
         conn.setAutoCommit(false) // Everything in the same db transaction.
         conn.setTransactionIsolation(finalIsolationLevel)
       }
-      val stmt = insertStatement(conn, table, rddSchema, dialect)
-      val setters: Array[JDBCValueSetter] = rddSchema.fields.map(_.dataType)
+      val stmt = insertStatement(conn, table, schema, dialect)
+      val setters: Array[JDBCValueSetter] = schema.fields.map(_.dataType)
         .map(makeSetter(conn, dialect, _)).toArray
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Could you just simplify it? 
```Scala
val setters = schema.fields.map(f => makeSetter(conn, dialect, f.dataType))
```

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Sure!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 12 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: After rethinking about it, the result of this function `normalizeSchema` is a mixed of two schemas (source/rdd schema + target/table schema). We should not mix them together.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 15]
```diff
@@ -568,10 +617,9 @@ object JdbcUtils extends Logging {
         conn.setAutoCommit(false) // Everything in the same db transaction.
         conn.setTransactionIsolation(finalIsolationLevel)
       }
-      val stmt = insertStatement(conn, table, rddSchema, dialect)
-      val setters: Array[JDBCValueSetter] = rddSchema.fields.map(_.dataType)
-        .map(makeSetter(conn, dialect, _)).toArray
-      val numFields = rddSchema.fields.length
+      val stmt = insertStatement(conn, table, schema, dialect)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `insertStatement` should use the target/table schema; the other places in this function should use the source/rdd schema.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 12 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Then, do you want to create `rddSchema` StructField with tableSchema's column name (if not matched)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 15 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: See the comment in https://github.com/apache/spark/pull/15664#discussion_r94094625

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I see.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Ur, `target/table schema` has a different column order than `rdd` order. Why do we need to use `table` schema? Actually, we need table column names only.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: We need to think about what is a better API interface of `savePartition`. Let me submit another commit and show you my rough idea.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 16]
```diff
@@ -211,6 +211,55 @@ object JdbcUtils extends Logging {
...
   /**
+   * Returns the schema if the table already exists in the JDBC database.
+   */
+  def getSchema(conn: Connection, url: String, table: String): Option[StructType] = {
+    val dialect = JdbcDialects.get(url)
+
+    try {
+      val statement = conn.prepareStatement(dialect.getSchemaQuery(table))
+      try {
+        Some(getSchema(statement.executeQuery(), dialect))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `getSchema` will throw an exception when the schema contains an unsupported type. Now we use it to check if the table exists. Does it change current behavior? E.g., the insertion working before now fails.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 15 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: This is my rough idea: https://github.com/apache/spark/compare/master...gatorsmile:pr-15664Changed1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 17]
```diff
@@ -211,6 +211,55 @@ object JdbcUtils extends Logging {
...
+    var schema = new StructType()
+    rddSchema.fields.foreach { f =>
+      if (nameMap.isDefinedAt(f.name)) {
+        // identical names
+        schema = schema.add(nameMap(f.name))
+      } else if (!caseSensitive && lowercaseNameMap.isDefinedAt(f.name.toLowerCase)) {
+        // identical names in a case-insensitive way
+        schema = schema.add(lowercaseNameMap(f.name.toLowerCase))
+      } else {
+        throw new AnalysisException(s"""Column "${f.name}" not found""")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: The error message looks ambiguous. Maybe `Column "${f.name}" in RDD not found in table schema`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 16 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: For unsupported types, it will throw an `SQLException`, right?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: yes.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Then, we can keep the existing way. See https://github.com/apache/spark/compare/master...gatorsmile:pr-15664Changed1

**Comment by dongjoon-hyun** [comment]
Thank you for deep review again, @viirya and @gatorsmile .
I'll update the PR soon.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 15 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yep. That looks much better than the current! Thank you.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 17 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: It may affect some other test cases, but I'll try!

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Oh, it's already in @gatorsmile 's patch.
```
        throw new AnalysisException(s"""Column "${col.name}" not found in schema $tableSchema""")
```

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile .
I have no objections on your patch! So, I merged that into this PR.

**Comment by dongjoon-hyun** [comment]
The only failure is irrelevant to this PR.
```
[info] StreamSuite:
[info] - fatal errors from a source should be sent to the user *** FAILED *** (101 milliseconds)
[info]   org.apache.spark.sql.streaming.StreamingQueryException: Query [id = b3869a67-68d5-4fc1-a3e2-d5879a49166c, runId = cdb70502-63b9-439a-9ece-bdbdcad5c2b4] terminated with exception: null
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runBatches(StreamExecution.scala:296)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:186)
[info]   Cause: org.apache.spark.sql.streaming.StreamSuite$$anonfun$12$$anon$2:
```

**Comment by dongjoon-hyun** [comment]
Retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 18]
```diff
@@ -57,26 +57,28 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
       val tableExists = JdbcUtils.tableExists(conn, url, table)
       if (tableExists) {
+        val tableSchema = JdbcUtils.getSchemaOption(conn, url, table)
         mode match {
           case SaveMode.Overwrite =>
             if (isTruncate && isCascadingTruncateTable(url) == Some(false)) {
               // In this case, we should truncate table and then load.
               truncateTable(conn, table)
-              saveTable(df, url, table, jdbcOptions)
+              saveTable(df, url, table, tableSchema.get, isCaseSensitive, jdbcOptions)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: We need to pass `tableSchema`. It is not safe to do `tableSchema.get`. : )

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 19]
```diff
@@ -108,14 +108,32 @@ object JdbcUtils extends Logging {
...
-   * Returns a PreparedStatement that inserts a row into table via conn.
+   * Returns an Insert SQL statement for inserting a row into the target table via JDBC conn.
    */
-  def insertStatement(conn: Connection, table: String, rddSchema: StructType, dialect: JdbcDialect)
-      : PreparedStatement = {
-    val columns = rddSchema.fields.map(x => dialect.quoteIdentifier(x.name)).mkString(",")
+  def getInsertStatement(
+      table: String,
+      rddSchema: StructType,
+      tableSchema: StructType,
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: If `tableSchema` is None, we should follow the original way. Please add the extra logics. Thanks!

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Right. I see. I'll update like that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 18 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Sure! Now, there is no pre-check.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 20]
```diff
@@ -67,16 +68,18 @@ class JdbcRelationProvider extends CreatableRelationProvider
             if (isTruncate && isCascadingTruncateTable(url) == Some(false)) {
               // In this case, we should truncate table and then load.
               truncateTable(conn, table)
-              saveTable(df, url, table, jdbcOptions)
+              val tableSchema = JdbcUtils.getSchemaOption(conn, url, table)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: I moved this into `case` statements.
Since `JdbcUtils.tableExists` is used, `getSchemaOption` can be skipped for the other `SaveMode`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 21]
```diff
@@ -108,14 +108,36 @@ object JdbcUtils extends Logging {
...
-      : PreparedStatement = {
-    val columns = rddSchema.fields.map(x => dialect.quoteIdentifier(x.name)).mkString(",")
+  def getInsertStatement(
+      table: String,
+      rddSchema: StructType,
+      tableSchema: Option[StructType],
+      isCaseSensitive: Boolean,
+      dialect: JdbcDialect): String = {
+    val columns = if (tableSchema.isEmpty) {
+      rddSchema.fields.map(x => dialect.quoteIdentifier(x.name)).mkString(",")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: The legacy behavior is used when `tableSchema` is `None`.

**Comment by dongjoon-hyun** [comment]
The only failure is irrelevant to this PR.
```
[info] StreamSuite:
[info] - fatal errors from a source should be sent to the user *** FAILED *** (84 milliseconds)
[info]   org.apache.spark.sql.streaming.StreamingQueryException: Query [id = d522aafe-085e-43e4-b796-037695dec113, runId = fc2f679a-907a-4445-8042-9198649bb55d] terminated with exception: null
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runBatches(StreamExecution.scala:296)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:186)
[info]   Cause: org.apache.spark.sql.streaming.StreamSuite$$anonfun$12$$anon$2:
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Merging to master. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile .
Happy New Year! :)
